### ISSUE TITLE
"this:" in @typedef fails to suppress `JSC_USED_GLOBAL_THIS` warnings in functions declared with @type

### ISSUE THREAD

**Issue by cpcallen** [google/closure-compiler#2857]
When declaring functions using @type {TypedefedName}, as in the following example:

```JS
/** @constructor */
function Foo() {
  this.bar = 42;
}

/** @typedef {function(this: Foo): number} */
Foo.funcType;

/** @const {Object<string, Foo.funcType>} */
var funcs = {};

/** @type {Foo.funcType} */
funcs['baz'] = function () {
  return this.bar;
}
```
…I get an unexpected warning:
```
Number of warnings: 1

JSC_USED_GLOBAL_THIS: dangerous use of the global this object at line 20 character 9
  return this.bar;
         ^
```
In `baz`, the compiler seems to understand that `this` is a `Foo`, because it will complain if I attempt to assign a string to `this.bar`, but it still emits the global this warning anyway.

(I can make the warning go away by explicitly declaring `baz` like this:
```JS
/**
 * @this {!Foo}
 * @return {number}
 */
funcs['baz'] = function () {
  return this.bar;
}
```
…but I have many such baz to define, each of which, in reality, has a long list of parameters, and it would be preferable not to repeat the full type declaration for each one.)

This occurs in v20180204 as well as [the current version on closure-compiler.appspot.com](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F%252F%2520ADD%2520YOUR%2520CODE%2520HERE%250A%250A%252F**%250A%2520*%2520%2540constructor%250A%2520*%252F%250Afunction%2520Foo()%2520%257B%250A%2520%2520this.bar%2520%253D%252042%253B%250A%257D%250A%250A%252F**%250A%2520*%2520%2540typedef%2520%257Bfunction(this%253A%2520Foo)%253A%2520number%257D%250A%2520*%252F%250AFoo.funcType%253B%250A%250A%252F**%2520%2540const%2520%257BObject%253Cstring%252C%2520Foo.funcType%253E%257D%2520*%252F%250Avar%2520funcs%2520%253D%2520%257B%257D%253B%250A%250A%252F**%2520%2540type%2520%257BFoo.funcType%257D%2520*%252F%250Afuncs%255B'baz'%255D%2520%253D%2520function%2520()%2520%257B%250A%2520%2520return%2520this.bar%253B%250A%257D%250A) when **advanced optimisations** are enabled.

**Comment by brad4d** [comment]
Looks to me like the type checker is just failing to look into the typedef, because no error is generated for [this version](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F%252F%2520ADD%2520YOUR%2520CODE%2520HERE%250A%250A%252F**%250A%2520*%2520%2540constructor%250A%2520*%252F%250Afunction%2520Foo()%2520%257B%250A%2520%2520this.bar%2520%253D%252042%253B%250A%257D%250A%250A%252F**%250A%2520*%2520%2540typedef%2520%257Bfunction(this%253A%2520Foo)%253A%2520number%257D%250A%2520*%252F%250AFoo.funcType%253B%250A%250A%252F**%2520%2540const%2520%257BObject%253Cstring%252C%2520Foo.funcType%253E%257D%2520*%252F%250Avar%2520funcs%2520%253D%2520%257B%257D%253B%250A%250A%252F%252F%252F**%2520%2540type%2520%257BFoo.funcType%257D%2520*%252F%250A%252F**%250A%2520*%2520%2540this%2520%257B!Foo%257D%250A%2520*%2520%2540return%2520%257Bnumber%257D%250A%2520*%252F%250Afuncs%255B'baz'%255D%2520%253D%2520function%2520()%2520%257B%250A%2520%2520return%2520this.bar%253B%250A%257D%250A).

```javascript
// ==ClosureCompiler==
// @compilation_level ADVANCED_OPTIMIZATIONS
// @output_file_name default.js
// ==/ClosureCompiler==

// ADD YOUR CODE HERE

/**
 * @constructor
 */
function Foo() {
  this.bar = 42;
}

/**
 * @typedef {function(this: Foo): number}
 */
Foo.funcType;

/** @const {Object<string, Foo.funcType>} */
var funcs = {};

///** @type {Foo.funcType} */
/**
 * @this {!Foo}
 * @return {number}
 */
funcs['baz'] = function () {
  return this.bar;
}
```

**Comment by brad4d** [comment]
@shicks,  do you think this would be an easy "oh, just look into the typedef definition here, we forgot to do that" kind of fix?

**Comment by cpcallen** [comment]
> Looks to me like the type checker is just failing to look into the typedef…

Yeah, that was what I thought initially too, but then @blickly  pointed out that the [compiler (correctly) complains about the type mismatch in this example](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F%252F%2520ADD%2520YOUR%2520CODE%2520HERE%250A%250A%252F**%250A%2520*%2520%2540constructor%250A%2520*%252F%250Afunction%2520Foo()%2520%257B%250A%2520%2520this.bar%2520%253D%252042%253B%250A%257D%250A%250A%252F**%250A%2520*%2520%2540typedef%2520%257Bfunction(this%253A%2520Foo)%257D%250A%2520*%252F%250AFoo.funcType%253B%250A%250A%252F**%2520%2540const%2520%257BObject%253Cstring%252C%2520Foo.funcType%253E%257D%2520*%252F%250Avar%2520funcs%2520%253D%2520%257B%257D%253B%250A%250A%252F**%2520%2540type%2520%257BFoo.funcType%257D%2520*%252F%250Afuncs%255B'baz'%255D%2520%253D%2520function%2520()%2520%257B%250A%2520%2520%2520var%2520%252F**%2520string%2520*%252F%2520s%2520%253D%2520this.bar%253B%250A%257D%250A):

```JS
/**
 * @constructor
 */
function Foo() {
  this.bar = 42;
}

/**
 * @typedef {function(this: Foo)}
 */
Foo.funcType;

/** @const {Object<string, Foo.funcType>} */
var funcs = {};

/** @type {Foo.funcType} */
funcs['baz'] = function () {
   var /** string */ s = this.bar;
}
```

```
JSC_TYPE_MISMATCH: initializing variable
found   : number
required: string at line 20 character 25
   var /** string */ s = this.bar;
                         ^
```
…so it's definitely looking into the typedef for *some* of the types.

**Comment by lauraharker** [comment]
I think the problem here is that the JSC_USED_GLOBAL_WARNING is emitted before typechecking, not during typechecking.

It's in the [CheckGlobalThis](https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/CheckGlobalThis.java) pass, which runs without any type information. The pass doesn't try to resolve what type `Foo.funcType` actually is.
